src/org/traccar/protocol/SviasProtocolDecoder.java

```diff
@@ -0,0 +1,150 @@
+
+    private double convertCoordinates(long v) {
+        return Double.valueOf(((float) ((((float) v / 1.0E7F)
+                - ((int) (v / 10000000L))) * 1.6666666666666667D)) + ((int) (v / 10000000L)));
+    }
+
+    @Override
+    protected Object decode(Channel channel, SocketAddress remoteAddress, Object msg)
+            throws Exception {
+
+        String sentence = (String) msg;
+
+        if (!sentence.contains(":")) {
+
+            Parser parser = new Parser(PATTERN, (String) sentence);
+            if (!parser.matches()) {
+                return null;
+            }
+
+            Position position = new Position(getProtocolName());
+
+            String versionHard = parser.next();
+            String versionSoft = parser.next();
+            String counterInternal = parser.next();
+
+            DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, parser.next());
+            if (deviceSession == null) {
+                return null;
+            }
+
+            position.setDeviceId(deviceSession.getDeviceId());
+
+            DateFormat dateFormat = new SimpleDateFormat("ddMMyyHHmmss");
+            dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+
+            String date = String.format("%06d", parser.nextInt());
+            String time = String.format("%06d", parser.nextInt());
+
+            position.setTime(dateFormat.parse(date + time));
+
+            position.setLatitude(convertCoordinates(parser.nextLong()));
+            position.setLongitude(convertCoordinates(parser.nextLong()));
+            position.setSpeed(UnitsConverter.knotsFromKph(parser.nextInt() / 100));
+            position.setCourse(parser.nextInt() / 100);
+            position.setAltitude(0);
+
+            position.set(Position.KEY_ODOMETER, parser.nextInt());
+
+            String input = new StringBuilder(String.format("%08d",
+                    Integer.parseInt(Integer.toString(parser.nextInt(), 2)))).reverse().toString();
+
+            String output = new StringBuilder(String.format("%08d",
+                    Integer.parseInt(Integer.toString(parser.nextInt(), 2)))).reverse().toString();
+
+            position.set(Position.KEY_ALARM, (input.substring(0, 1).equals("1")
+                    ? Position.ALARM_SOS : null));
+
+            position.set(Position.KEY_IGNITION, input.substring(4, 5).equals("1"));
+
+            position.setValid(output.substring(0, 1).equals("1"));
+
+            String pck1 = parser.next();
+            String pck2 = parser.next();
+
+            position.set(Position.KEY_POWER, parser.nextInt() / 1000);
+
+            position.set(Position.KEY_BATTERY_LEVEL, parser.nextInt());
+
+            position.set(Position.KEY_RSSI, parser.nextInt());
+
+            if (channel != null) {
+                channel.write(MSG_KEEPALIVE);
+            }
+
+            return position;
+
+        } else {
+            //send keepalive for message check
+            if (channel != null) {

```

Comment:
I think this can be done in a single place.

Reply:
Yes! Done!